tensorflow_probability/python/experimental/mcmc/with_reductions_test.py

```diff
@@ -0,0 +1,499 @@
+# Copyright 2020 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Tests for WithReductions TransitionKernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+# Dependency imports
+from absl.testing import parameterized
+import numpy as np
+
+import tensorflow.compat.v2 as tf
+import tensorflow_probability as tfp
+from tensorflow_probability.python.experimental.mcmc import CovarianceReducer
+from tensorflow_probability.python.experimental.mcmc import VarianceReducer
+from tensorflow_probability.python.experimental.mcmc import WithReductions
+from tensorflow_probability.python.internal import test_util
+
+
+class TestReducer(tfp.experimental.mcmc.Reducer):
+  """Simple Reducer that just keeps track of the last sample"""
+
+  def initialize(self, initial_chain_state, initial_kernel_results=None):
+    return tf.zeros(tf.convert_to_tensor(initial_chain_state).shape)
+
+  def one_step(
+      self, new_chain_state, current_reducer_state, previous_kernel_results):
+    print(new_chain_state)
+    return new_chain_state
+
+
+class MeanReducer(tfp.experimental.mcmc.Reducer):
+  """Simple Reducer that (naively) computes the mean"""
+
+  def initialize(self, initial_chain_state=None, initial_kernel_results=None):
+    return tf.zeros((2,))
+
+  def one_step(
+      self, new_chain_state, current_reducer_state, previous_kernel_results):
+    return current_reducer_state + tf.convert_to_tensor([1, new_chain_state])
```

Comment:
I'd use `tf.stack` to be explicit.

Reply:
I'm using the addition to auto-broadcast in the nested reducer state

Reply:
I meant `tf.stack([1, new_chain_state])` rather than `tf.convert_to_tensor`. I'm not actually seeing where the broadcasting happens, but I wasn't talking about that either way.

Reply:
Ohhhh oops, I understand now.